File: Design and Verification Training Program/Veriest/Scripting/Perl/create_module_syntax.pl
#!/usr/bin/perl -w
use strict;
use Data::Dumper;

my $file_name = $ARGV[0];
my $module_name = $ARGV[1];
my $module_param_name = $ARGV[2];
my $module_param_value = $ARGV[3];
open (FH, '<', $file_name) or die "Couldn't open file";
my @lines = <FH>; # load lines from text file
#print "@lines \n";
my @temp;
@temp = split(/[=]+/, $lines[0]);
#print "@temp \n";
$temp[1] =~ s/\s+//g; # remove all white spaces of the temp[1] string (dict)
$temp[1] =~ s/'+//g; # remove all " ' " of the temp[1] string (dict)
#print "$temp[1] \n";
substr($temp[1], 0, 1, ''); # remove first char ({)
substr($temp[1], -1, 1, ''); # remove last char (})
#print "$temp[1] \n";

@lines = split(/[{ \s\ }]+/, $temp[1]); # split string by "{" and "}"
#print "@lines \n";
my $index;
for ($index = 1; $index <= $#lines; $index = $index + 4){
    splice @lines, $index, 0, '{'; # insert "{" as array element at the "{" split points of the string
    splice @lines, $index + 2, 0, '}';  # insert "}" as array element at the "}" split points of the string
}
#print "@lines \n";

my @new_lines;
my $item;
foreach $item (@lines){
    push @new_lines, split(/[, \s\ :]+/, $item); # split strings in lines array by "," and ":"
}
#print "@new_lines \n\n";
@new_lines = grep /\S/, @new_lines; # remove empty elements from new_lines array
#print "@new_lines \n\n";

my %signals_hash;
for ($index = 0; $index <= $#new_lines; $index++){
    if ($new_lines[$index] eq '{'){
	my %signals;
	my $i = 1;
	while ($new_lines[$index + $i] ne '}'){
	    $signals{$new_lines[$index + $i]} = $new_lines[$index + $i + 1];
	    $i = $i + 2;
	}
	$signals_hash{$new_lines[$index - 1]} = \%signals;
    }
}
#print Dumper(\%signals_hash);

my $key;
my $k;
foreach $key (keys %signals_hash){
    foreach $k (keys %{$signals_hash{$key}}){
	if ($signals_hash{$key}{$k} eq "Null"){
	    delete $signals_hash{$key}{$k}; # remove "Null" values from hash
	}
    }
}
#print Dumper(\%signals_hash);

my $file = $module_name . ".sv";
open (FILE, '>'.$file) or die "Couldn't open file";
print FILE "module " . $module_name . "\n";
print FILE "\t" . "#(parameter " . $module_param_name . " = " . $module_param_value . ")\n";
print FILE "\t(\n";
foreach $key (keys %signals_hash){
    foreach $k (sort keys %{$signals_hash{$key}}){
	if (($key eq (keys %signals_hash)[-1]) && ($k eq (keys %{$signals_hash{$key}}))[-1]){ # if the current item is the very last item in the nested hashes
	    print FILE "\t$key\t$signals_hash{$key}{$k}\n";
	} else {
	    print FILE "\t$key\t$signals_hash{$key}{$k},\n";
	}
    }
}
print FILE "\t);\n";
print FILE "\n\nendmodule";
close FILE;

open (FH, $file) or die "Couldn't open file";
while (<FH>)
{
	print $_;
}
close;

File: Design and Verification Training Program/Veriest/Scripting/Python/create_module_syntax.py
#!/usr/bin/python
import sys
import re

file_name = sys.argv[1]
module_name = sys.argv[2]
module_param_name = sys.argv[3]
module_param_value = sys.argv[4]
signals = open(file_name,'r')
lines = signals.read()
#print(lines)

# Works perfectly, but bad practice:
#context = {}
#d = exec(lines, context)

#
# Convert dictionary given as string to dictionary type:
#
# Remove "d =" from the string read from the text file:
string_dict = lines.split("=")
#print(string_dict)
string_dict = string_dict[1]
#print(string_dict)
# Remove " ", "'" and the outer "{" "}" from the string:
string_dict = string_dict.replace(" ","")
string_dict = string_dict.replace("'","")
string_dict = string_dict[1:-1]
#print(string_dict)

# Split string_dict at "{":
lbraces = string_dict.split("{")
#print(lbraces)
# Insert ":{" into lbraces where it's supposed to be at:
for i in range (1,(len(lbraces)+len(lbraces)-1),2):
    lbraces.insert(i,"{")
#print(lbraces)

# Split lbraces items at "}":
rbraces = []
for j in lbraces:
    rbraces.append(j.split("}"))
#print(rbraces)
# Insert ":{" into rbraces where it's supposed to be at:
for k in rbraces:
    for l in range (1,(len(k)),2):
        k.insert(l,'}')
#print(rbraces)

# Split items of rbraces at "," and ":":
comma_colon = []
for a in rbraces:
    for b in a:
        comma_colon.append(re.split(',|:',b))

# Flatten comma_colon list (all nested items become non-nested items in flat_list):
flat_list = []
for sublist in comma_colon:
    for item in sublist:
        flat_list.append(item)
#print(flat_list)

# Remove blank items from flat_list
while '' in flat_list:
    flat_list.remove('')
#print(flat_list)

d = {}
for index in range (0,len(flat_list)):
    if flat_list[index] == '{':
        my_dict = {}
        i = 1
        #print(index)
        while flat_list[index+i] != '}':
            my_dict[flat_list[index+i]] = flat_list[index+i+1]
            i = i + 2
        d[flat_list[index-1]] = my_dict
#print(d)

#d = {'input': {'1': 'haddr', '2': 'hwrite', '3': 'hsize' , '4': 'hsel', '5': 'hburst', '6': 'hprot', '7': 'htrans', '8': 'hmastlock', '8': 'hready', '9': 'hmastlock', '10': 'hwdata', '11' : 'hresetn', '12': 'hclk', '13': 'Null', '14': 'sys_clk', '15': 'sys_reset', '16': 'sys_err'},'output': {'1': 'hrdata', '2': 'hresp','3': 'hreadyout'},'inout': {'1': 'datap', '2': 'datan','3': 'clkp', '4': 'clkn'}}

# Remove dict items with "Null" values:
delete_list = []
for k in d.keys():
    for i_k in d[k].keys():
        if d[k][i_k] == "Null":
            delete_list.append([k,i_k])
for [x,y] in delete_list:
    del d[x][y] 

#
# Create sv module template:
#
f = open(module_name+'.sv','w')
f.write("module " + module_name + "\n")
f.write("\t" + "#(parameter " + module_param_name + " = " + module_param_value + ")\n")
f.write("\t(\n")

for key in d:
    for inner_key in d[key]:
        if ((key == d.keys()[-1]) & (inner_key == d[key].keys()[-1])): # if the current item is the very last item in the nested dicts
            f.write("\t" + key + "\t" + d[key][inner_key] + "\n") 
        else:
            f.write("\t" + key + "\t" + d[key][inner_key] + ",\n")

f.write("\t);\n")
f.write("\n\nendmodule")
f.close()
f = open(module_name+'.sv','r')
print(f.read())
